tiledb/sm/query/reader_base.cc

```diff
@@ -414,228 +414,200 @@ Status ReaderBase::read_tiles(
   if (result_tiles->empty())
     return Status::Ok();
 
-  // Reading tiles are thread safe. However, we will perform
-  // them on this thread if there is only one read to perform.
-  if (names->size() == 1) {
-    RETURN_NOT_OK(read_tiles(names->at(0), result_tiles));
-  } else {
-    const auto status = parallel_for(
-        storage_manager_->compute_tp(),
-        0,
-        names->size(),
-        [&](const uint64_t i) {
-          RETURN_NOT_OK(read_tiles(names->at(i), result_tiles));
-          return Status::Ok();
-        });
-
-    RETURN_NOT_OK(status);
-  }
-
-  return Status::Ok();
-}
-
-Status ReaderBase::read_tiles(
-    const std::string& name,
-    const std::vector<ResultTile*>* result_tiles) const {
-  // Shortcut for empty tile vec
-  if (result_tiles->empty())
-    return Status::Ok();
-
-  // Read the tiles asynchronously
-  std::vector<ThreadPool::Task> tasks;
-  RETURN_CANCEL_OR_ERROR(read_tiles(name, result_tiles, &tasks));
-
-  // Wait for the reads to finish and check statuses.
-  auto statuses = storage_manager_->io_tp()->wait_all_status(tasks);
-  for (const auto& st : statuses)
-    RETURN_CANCEL_OR_ERROR(st);
-
-  return Status::Ok();
-}
-
-Status ReaderBase::read_tiles(
-    const std::string& name,
-    const std::vector<ResultTile*>* result_tiles,
-    std::vector<ThreadPool::Task>* const tasks) const {
-  // Shortcut for empty tile vec
-  if (result_tiles->empty())
-    return Status::Ok();
-
-  // For each tile, read from its fragment.
-  const bool var_size = array_schema_->var_size(name);
-  const bool nullable = array_schema_->is_nullable(name);
+  // Populate the list of regions per file to be read.
+  std::map<URI, std::vector<std::tuple<uint64_t, void*, uint64_t>>> all_regions;
+  std::mutex all_regions_mutex;
 
-  // Gather the unique fragments indexes for which there are tiles
-  std::unordered_set<uint32_t> fragment_idxs_set;
-  for (const auto& tile : *result_tiles)
-    fragment_idxs_set.emplace(tile->frag_idx());
+  // Mutex protecting attr_tiles_ maps in all ResultTiles.
```

Comment:
Instead of using an external lock on the data structure should we instead add the mutex inside `tile`?  Also is this a place where a concurrent map/unordered_map would simplify things? cc: @lums658

Reply:
I don't think we want to use a concurrent map as this will make everywhere in the code that uses it, including places where performance might be an issue impacted. I first tried to put the lock in the result tile, but since having an extra mutex would add 40 bytes to a structure that's 500 bytes that we create a lot of that get moved often, I decided against it.

Reply:
I'd love to get @eric-hughes-tiledb and @lums658 feedback here on size of the object vs the benefits of keeping the data structure self contained.

I don't think we need to change anything specifically in this PR but this is exactly the situation we are trying to rework in order to avoid the complex access and lifetime of objects we have today.

Reply:
@Shelnutt2 Of possible consideration, #s of mutexes to be defined (# of items potentially involved?) vs any os required resources (for those mutexes) available?
(I thought once upon a time that visual studio cpp std::mutex had an os handle dependency, 
but appears that is -not- currently the case.)